### ISSUE TITLE
cors issue with uploading to signed url

### ISSUE THREAD

**Issue by joshyg** [googleapis/nodejs-storage#347]
I have tried many configurations to get file upload with signed url.  preflight request completes ok but actual upload fails with 
No 'Access-Control-Allow-Origin' header is present on the requested resource.

This makes me thin that the cors config is ok but there is an issue with the way the url is being signed. 
I see a lot of people having issues with this on stack overflow but no resolutions, which makes me think that this is really a documentation issue. Below is the most lenient configuration I have tried:

cors config:

```js
[{"maxAgeSeconds": 3600, "method": ["GET", "HEAD", "POST", "PUT"], "origin": ["*"], "responseHeader": ["Content-Type", "Access-Control-Allow-Origin"]}]
```

serverside code:

```js
   app.post('/api/v1/upload/gcs-sign', function(req, res) {
     const fileName = req.body.fileName;
     const fileType = req.body.fileType;
     const file = bucket.file(fileName);

     // signed URL expires in 1 hour
     const expirationDate = new Date(
        new Date().setHours(new Date().getHours() + 1),
    );
     const config = {
        action: 'resumable',
        expires: expirationDate,
        contentType: req.body.fileType,
     };
     file.getSignedUrl(config, (err, signedRequest) => {
        if (err) {
          return res.end();
        }
       res.json({ signedRequest });
      res.end();
    })
```

client side code:

```js
     uploadToGCS = (signedRequest, file) => {
      let xhr = new XMLHttpRequest();
      xhr.open('POST', signedRequest);
      xhr.onload = () => {
        this.onUploadFinish(xhr);
      };
      xhr.onerror = this.onUploadError;
     xhr.upload.onprogress = this.onUploadProgress;
     xhr.setRequestHeader('Content-Type', file.type);
     xhr.send(file);
   };
```

#### Environment details

  - OS: OSX 10.13.3
  - Node.js version: 10.7.0
  - npm version: 6.2.0
  - `@google-cloud/storage` version: 1.7.0

#### Steps to reproduce

  1. create bucket witth cors config above
  2. run server code above ( Ive left out some basic boiler plate )
  3.  attempt to upload file with client side code above

Thanks!

**Comment by joshyg** [comment]
Follow up:

to be more clear, this issue is really a development issue.  I am running a node server from localhost:5000 for my backend, and then serving a react app from localhost:8080.  If I serve the pre built react app from the node server then everything is fine, but if I want to run a development server, which does hot reloads and therefore makes life much easier for me, I just cant get the CORS issues to go away.  I can manually set an origin for resumable url creation, but not for signed urls.

**Comment by yeya** [comment]
An optional workaround is to setup the HMR to work with the node server.
With vue.js it can be done very easily in 3 simple steps as you can read [here](https://code.luasoftware.com/tutorials/vuejs/enable-vuejs-webpack-development-with-hot-module-reloading-on-external-server/)

I guess it can be done in a similar way for React too.

**Comment by fhinkel** [comment]
@joshyg Thanks for reporting this. We just released Storage 2.0. Would you mind checking if this is still an issue? 🙏

**Comment by mrdulin** [comment]
`"@google-cloud/storage": "^2.5.0",`

Got 403 CORS issue

**Comment by sebght** [comment]
I do have the same issue as you do @mrdulin 
For more details about mine :
Preflight `OPTIONS` request responds ok with the correct headers, and then my `PUT` request got 403 too

**Comment by mrdulin** [comment]
@sebght try change `version` option for `getSignedUrl` from `v2`(default) to `v4`

**Comment by sebght** [comment]
Quite surprisingly, adding `"Content-Type": file.type` removed the CORS issue with my case 😄

**Comment by lanuma** [comment]
Hi, sorry for reopen this issue, i had same issue.
In my request i have to add `Content-Type` header when PUT method, but the cors still appear.

can you help me?

**Comment by JustinBeckwith** [comment]
Greetings!  If you're having issues, would you mind opening a new issue?  It can be hard to find and track these when they're closed :)

**Comment by caseywebdev** [comment]
> @sebght Try change `version` option for `getSignedUrl` method from `v2`(default) to `v4`

This made the difference for me. There's clearly an issue with v2 and CORS. As soon as I switched to v4 the same CORS settings worked fine.

**Comment by gabriellanzer** [comment]
When I try to use v4 I get an unknown error! How do I change it? Getting 403 even with CORS file defined with: ["GET", "POST", "PUT", "DELETE", "OPTIONS"], "origin": ["*"], "responseHeader": ["Content-Type", "Authorization", "Content-Length", "User-Agent", "x-goog-resumable", "Access-Control-Allow-Origin"]}]...

**Comment by gabriellanzer** [comment]
Now it works... I was using POST instead of PUT and I had to set the Content-Type on both the getSignedURL config and the client's POST request header.

**Comment by gabriellanzer** [comment]
If someone happens to fall into this Closed Issue, chrome was giving me 403 but the response had a message saying "The request signature we calculated does not match the signature you provided.", I suggest using Postman to improve debugging...

**Comment by WolverineRV** [comment]
> Now it works... I was using POST instead of PUT and I had to set the Content-Type on both the getSignedURL config and the client's POST request header. I could get it working using Content-Type to 'application/octet-stream'.

THANK YOU!!!!  Fought with this for DAYS.  Setting the content type to application/octet-stream in both parts did the trick!

**Comment by leonsomed** [comment]
I was using fetch with FormData to send the file, for whatever reason that did not work. I ended up doing a couple of things:

- Adding CORS configuration to the bucket as described here https://cloud.google.com/storage/docs/configuring-cors
- Switching from fetch to XMLHttpRequest to send the file
- Using PUT instead of POST
- Using Content-Type: application/octet-stream in both the server and the client

After doing all of those I finally have it working.

**Comment by MOIN-AKHTAR** [comment]
Getting error when sending post request using signedUrl for resumable upload from front end.Getting CORS issue not getting sessionUrl


Hello everyone i'm using gcp bucket for resumable upload i have created an endpoint on backend whihc will give me a signedUrl.
BACKEND CODE:-
```
async getSignedUrl(query, res) {
    try {
      res.set('Access-Control-Allow-Origin', 'http://localhost:3000/');
      const [url] = await storage
        .bucket('my-bucket')
        .file(query.fileName)
        .getSignedUrl({
          action: 'resumable',
          version: 'v4',
          expires: Date.now() + 12 * 60 * 60 * 1000,
          contentType: 'application/octet-stream',
        });
      return res.json({
        url,
      });
    } catch (error) {
      throw new HttpException(
        error.message || error.response || error,
        HttpStatus.BAD_REQUEST,
      );
    }
  }
```
Now i want to use signedUrl given by this endpoint on front end for resumable upload.Now inorder to used it as resumable according to documetation we have to hit a post request using this signedUrl that will give us SessionUrl this session url will be used for resumable upload but when i'm hitting post request using signedUrl which i got from backed refer to above code i'm getting this error.
FRONT END CODE (REACTJS):-
```
const uploadChunk = useCallback(
    (readerEvent) => {
      axios({
        method: "POST",
        url, ===> Reffering signedUrl which we got from backend code .
      })
        .then((res) => {
          console.log(res.data);
        })
        .catch((err) => {
          console.log(err);
        });
    },
    [url]
  );
```
Access to XMLHttpRequest at 'https://storage.googleapis.com/blahblahbla...' from origin 'http://localhost:3000/' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
Than using gsutils my cors config is this:

`[{"maxAgeSeconds": 3600, "method": ["GET", "POST", "PUT", "DELETE", "HEAD", "OPTIONS"], "origin": ["http://localhost:3000/"], "responseHeader": ["Content-Type", "Authorization", "x-goog-resumable", "Access-Control-Allow-Origin"]}]
`
```
I'm using :
"@google-cloud/storage": "^6.1.0"
```

Than why i'm getting this error any idea please

**Comment by jennwah** [comment]
Google Cloud seriously needs to document this somewhere
